vizro-ai/docs/pages/user-guides/vizro-ai-langchain-guide.md

```diff
@@ -0,0 +1,193 @@
+import vizro.plotly.express as px
+from langchain_core.runnables import chain
+from langchain_core.tools import InjectedToolArg, tool
+from langchain_openai import ChatOpenAI
+from vizro_ai import VizroAI
+
+# Load sample data
+df = px.data.gapminder()
+dfs = [df]
+
+# Initialize the LLM
+llm = ChatOpenAI(model="gpt-4o-mini")
+```
+
+## 2. Define LangChain tools
+
+Basic tools only take string as input and output. Vizro-AI takes Pandas dataframes as input and it's not cost-efficient and secure to pass the full data to a LLM. In this case, [bind the Pandas dataframes at run time](https://python.langchain.com/v0.2/docs/how_to/tool_runtime/) is more suitable.
+
+Now, create tools that wrap Vizro-AI's plotting and dashboard generation capabilities:
+
+```python
+@tool(parse_docstring=True)
+def get_plot_code(df: Annotated[Any, InjectedToolArg], question: str) -> str:
+    """Generate only the plot code.
+
+    Args:
+        df: A pandas DataFrame
+        question: The plotting question
+
+    Returns:
+        Generated plot code
+    """
+    df = pd.DataFrame(df)
+    vizro_ai = VizroAI(model=llm)
+    plot_elements = vizro_ai.plot(
+        df,
+        user_input=question,
+        return_elements=True,
+    )
+    return plot_elements.code_vizro
+
+@tool(parse_docstring=True)
+def get_dashboard_code(dfs: Annotated[Any, InjectedToolArg], question: str) -> str:
+    """Generate the dashboard code.
+
+    Args:
+        dfs: Pandas DataFrames
+        question: The dashboard question
+
+    Returns:
+        Generated dashboard code
+    """
+    vizro_ai = VizroAI(model=llm)
+    dashboard_elements = vizro_ai.dashboard(
+        dfs,
+        user_input=question,
+        return_elements=True,
+    )
+    return dashboard_elements.code
+```
+
+## 3. Set up the tool chain
+
+Create a chain that handles tool execution and data injection:
+
+```python
+# Bind tools to the LLM
+tools = [get_plot_code, get_dashboard_code]
+llm_with_tools = llm.bind_tools(tools)
+
+# Create data injection chain
+@chain
+def inject_df(ai_msg):
+    tool_calls = []
+    for tool_call in ai_msg.tool_calls:
+        tool_call_copy = deepcopy(tool_call)
+
+        if tool_call_copy["name"] == "get_dashboard_code":
+            tool_call_copy["args"]["dfs"] = dfs
+        else:
+            tool_call_copy["args"]["df"] = df
+
+        tool_calls.append(tool_call_copy)
+    return tool_calls
+
+# Create tool router
+tool_map = {tool.name: tool for tool in tools}
+
+@chain
+def tool_router(tool_call):
+    return tool_map[tool_call["name"]]
+
+# Combine chains
+chain = llm_with_tools | inject_df | tool_router.map()
+```
+
+## 4. Use the chain
+
+Now you can use the chain to generate charts or dashboards based on natural language queries. The chain will generate code that you can use to create visualizations.
+
+!!! example "Generate chart code"
+
+    === "Code"
+        ```py
+        plot_response = chain.invoke("Plot GDP per capita for each continent")
```

Comment:
How would I invoke the chain with a different df, can that we passed as an argument, or does this only work if hard-coded at the top?

Reply:
I think the key is which data `df` is pointing to at run time of `chain.invoke()`. So we can do the following:
```
df = px.data.gapminder()
plot_response = chain.invoke("Plot GDP per capita for each continent")

df = px.data.iris()
plot_response = chain.invoke("Plot the distribution of petal length")
```